### PR TITLE
test: refactor testing tools

### PR DESCRIPTION
**Pull request by lemonholic**
- Check the vacancy of ports before assigning them to the test functions.
- Refactor some test helper functions to be more unified and reusable.

### CODE REVIEW

File: network/src/propagation_network/mod.rs [hunk 1]
```diff
@@ -384,6 +417,137 @@ mod test {
...
+        async fn add_nodes_to_network_sequential(
+            &mut self,
+            max_bootstrap_points: usize,
+        ) -> Result<Vec<NodeKey>, String> {
+            let target_nodes = self.get_nodes_not_in_network();
+
+            for key in &target_nodes {
+                // Select random nodes and add them to bootstrap points.
+                let mut bootstrap_points = Vec::new();
+                let mut rng = rand::thread_rng();
```
Comment (junha1) [inline network/src/propagation_network/mod.rs:468]: Can this cause a nondeterministic result of any test?
Unlike the randomness from port selection  or key generation (which wouldn't really matter), this seems quite effective.

File: network/src/propagation_network/mod.rs [hunk 2]
```diff
@@ -384,6 +417,137 @@ mod test {
...
+
+            // Create the network interfaces asynchronously.
+            let nodes: Vec<&Node> = target_nodes
+                .iter()
+                .map(|key| self.nodes.get(key).unwrap())
+                .collect();
+            let futures = zip(nodes, &listen_addresses).map(|(node, listen_address)| {
+                let mut config = create_testnet_config();
+                config.with_listen_address(listen_address.to_owned());
+                let mut rng = rand::thread_rng();
```
Comment (junha1) [inline network/src/propagation_network/mod.rs:523]: ditto.

File: network/src/propagation_network/mod.rs (hunk 1 above)
Comment (lemonholic) [reply network/src/propagation_network/mod.rs:468]: 1. If `max_bootstrap_points == total_nodes_in_network`, the network will be initialed with fully-connected nodes, so the process and the result of peer discovery are deterministic. 
2. If we select random bootstrap nodes less than the number of total nodes in the network, the process of peer discovery will be non-deterministic. However, the expected result should always be the same if our code works well; each node figures out and keeps 20 nodes by querying given bootstrap nodes.

**Comment by lemonholic** [comment]
Rebased on the most recent commit.

File: network/src/propagation_network/mod.rs (hunk 2 above)
Comment (lemonholic) [reply network/src/propagation_network/mod.rs:523]: check the comment above

**Review by junha1 — APPROVED** [review]

**Comment by lemonholic** [comment]
rebased

**Pull request merged**